metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java

```diff
@@ -0,0 +1,246 @@
+                setPartitionId(0).
+                setIsr(Arrays.asList(2, 3)).
+                setLeader(2), PARTITION_CHANGE_RECORD.highestSupportedVersion())),
+            createFooBuilder(false).setTargetIsr(Arrays.asList(2, 3)).build());
+    }
+
+    @Test
+    public void testReassignmentRearrangesReplicas() {
+        assertEquals(Optional.of(new ApiMessageAndVersion(new PartitionChangeRecord().
+                setTopicId(FOO_ID).
+                setPartitionId(0).
+                setReplicas(Arrays.asList(3, 2, 1)),
+                PARTITION_CHANGE_RECORD.highestSupportedVersion())),
+            createFooBuilder(false).setTargetReplicas(Arrays.asList(3, 2, 1)).build());
+    }
+
+    @Test
+    public void testIsrEnlargementCompletesReassignment() {
+        assertEquals(Optional.of(new ApiMessageAndVersion(new PartitionChangeRecord().
+                setTopicId(BAR_ID).
+                setPartitionId(0).
+                setReplicas(Arrays.asList(2, 3, 4)).
+                setIsr(Arrays.asList(2, 3, 4)).
+                setLeader(2).
+                setRemovingReplicas(Collections.emptyList()).
+                setAddingReplicas(Collections.emptyList()),
+                PARTITION_CHANGE_RECORD.highestSupportedVersion())),
+            createBarBuilder(false).setTargetIsr(Arrays.asList(1, 2, 3, 4)).build());
+    }
+
+    @Test
+    public void testRevertReassignment() {
+        PartitionReassignmentRevert revert = new PartitionReassignmentRevert(BAR);
+        assertEquals(Arrays.asList(1, 2, 3), revert.replicas());
+        assertEquals(Arrays.asList(1, 2, 3), revert.isr());
+        assertEquals(Optional.of(new ApiMessageAndVersion(new PartitionChangeRecord().
+                setTopicId(BAR_ID).
+                setPartitionId(0).
+                setReplicas(Arrays.asList(1, 2, 3)).
+                setLeader(1).
+                setRemovingReplicas(Collections.emptyList()).
+                setAddingReplicas(Collections.emptyList()),
+                PARTITION_CHANGE_RECORD.highestSupportedVersion())),
+            createBarBuilder(false).
+                setTargetReplicas(revert.replicas()).
+                setTargetIsr(revert.isr()).
+                setTargetRemoving(Collections.emptyList()).
+                setTargetAdding(Collections.emptyList()).
+                build());
+    }
+
+    @Test
+    public void testQuickReassignment() {
+        PartitionReassignmentReplicas replicas = new PartitionReassignmentReplicas(
+            Replicas.toList(FOO.replicas), Arrays.asList(1, 2));
+        assertEquals(Collections.singletonList(3), replicas.removing());
+        assertEquals(Collections.emptyList(), replicas.adding());
+        assertEquals(Arrays.asList(1, 2, 3), replicas.merged());
+        assertEquals(Optional.of(new ApiMessageAndVersion(new PartitionChangeRecord().
+                setTopicId(FOO_ID).
+                setPartitionId(0).
+                setReplicas(Arrays.asList(1, 2)).
```

Comment:
This reassignment completes immediately because it's a pure removal (nothing added) and the ISR is ready to go immediately.

(It would be more typical to also add a new node to replace the one being removed, but in this case we aren't)